node/lib/util/rm.js

```diff
@@ -0,0 +1,482 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert       = require("chai").assert;
+const binarySearch = require("binary-search");
+const co           = require("co");
+const fs           = require("fs-promise");
+const path         = require("path");
+const NodeGit      = require("nodegit");
+
+const CloseUtil           = require("./close_util");
+const StatusUtil          = require("./status_util");
+const SubmoduleUtil       = require("./submodule_util");
+const SubmoduleConfigUtil = require("./submodule_config_util");
+const TextUtil            = require("./text_util");
+const UserError           = require("./user_error");
+
+const FILEMODE = NodeGit.TreeEntry.FILEMODE;
+
+const compareIndexEntryToPath = function(indexEntry, path) {
+    return TextUtil.strcmp(indexEntry.path, path);
+};
+
+const needToRequestRecursive = function(path) {
+    throw new UserError(
+        `not removing '${path}' recursively without -r`);
+};
+
+/**
+ * helper for getRemovalSpecs
+ * @param first The first index to look at
+ */
+/**
+ * There are four possible things that a user might be trying to
+ * remove:
+ *
+ * 1. A file inside a submodule
+ * 2. A subdirectory inside a submodule
+ * 3. A submodule or file within the meta repo
+ * 4. A directory containing multiple such things
+ *
+ * In order to handle cases 2 and 4, we will delegate to nodegit2's
+ * index.removeAll.  We need to ensure that there is at least one
+ * object in the (meta or sub) index that matches the user's request;
+ * if there is not, we throw an exception.
+ *
+ * @visibleForTesting
+ *
+ * @async
+ * @param {String} pathToRemove Normalized to the root of the meta
+ * repo, and no trailing slashes.  In the event that the user desires
+ * to remove everything in the repo, the pathspec "." must be
+ * translated to "", which is illegal from the git CLI's perspective,
+ * but which is legit internally.
+ *
+ * @param {Boolean} recursive If false, and the path to remove is a directory,
+ * give the "can't remove a dir without -r" message
+ *
+ * @return {Object}
+ * @return {Boolean} return.isRootLevel -- distinguishes cases 1/2 from 3/4
+ * @return {String}  return.pathFromRoot to the submodule
+ * @return {String}  return.pathInSubmodule
+ * @return {NodeGit.Repo}  return.submoduleRepo
+ */
+exports.getRemovalSpec = co.wrap(function *(repo, index, pathToRemove,
+                                             recursive) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(index, NodeGit.Index);
+    assert.isString(pathToRemove);
+    assert.isBoolean(recursive);
+
+    const noSuchFile = function() {
+        throw new UserError(`pathspec '${pathToRemove}' did not match any \
+files`);
+    };
+
+    if (pathToRemove.endsWith("/")) {
+        pathToRemove = pathToRemove.substring(0, pathToRemove.length - 1);
+    }
+
+    const idx = binarySearch(index.entries(), pathToRemove,
```

Comment:
I don't see anything in the documentation to indicate that the array returned by `entries` is sorted. Am I missing something? If not, would `getByPath` work?

Reply:
(nvm the `getByPath` suggestion)

Reply:
The index is stored on disk sorted, and in regular git code, it is standard to assume that the in-memory representation is as well.  I don't see a reason why either libgit2 or nodegit would change the ordering, so I think this is probably safe despite the lack of doco.

Reply:
Makes sense.

Reply:
... though you might want to put a comment saying as much there